src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs

```diff
@@ -0,0 +1,175 @@
+            List<RequiredPaymentEntity> refunds,
+            List<HistoricalPaymentEntity> previousPayments)
+        {
+            var refundGroups = refunds.ToLookup(x => new RefundGroup(x));
+            var volatilePreviousPayments = new List<HistoricalPaymentEntity>(previousPayments);
+
+            var refundPayments = new List<PaymentEntity>();
+            
+            foreach (var refundGroup in refundGroups)
+            {
+                foreach (var refund in refundGroup)
+                {
+                    var previousPaymentGroup = volatilePreviousPayments.ToLookup(x => new RefundGroup(x));
+                    var previousPaymentsForRefundGroup = new List<HistoricalPaymentEntity>();
+                    try
+                    {
+                        if (previousPaymentGroup.Contains(refundGroup.Key))
+                        {
+                            previousPaymentsForRefundGroup = previousPaymentGroup[refundGroup.Key].ToList();
+                        }
+
+                        var newRefunds = ProcessRefund(refund, previousPaymentsForRefundGroup);
+                        refundPayments.AddRange(newRefunds);
+                        volatilePreviousPayments.AddRange(newRefunds.Select(x => new HistoricalPaymentEntity(x, refund)));
+                    }
+                    catch(ApplicationException)
+                    { 
+                        // A funding source is negative, so ignoring this refund
+                        _logger.Error($"Refund for UKPRN: {refund.Ukprn} " +
+                                      $"LearnRefNumber: {refund.LearnRefNumber} " +
+                                      $"Amount: {refund.AmountDue} " +
+                                      $"DeliveryMonth: {refund.DeliveryMonth} " +
+                                      $"not made due to a net negative funding source");
+                    }
+                }
+            }
+
+            return refundPayments;
+        }
+
+        private List<PaymentEntity> ProcessRefund(
+            RequiredPaymentEntity refund, 
+            List<HistoricalPaymentEntity> previousPayments)
+        {
+            var refundPayments = new List<PaymentEntity>();
+
+            var amountToRefund = refund.AmountDue;
+            var amountRefunded = 0m;
+            var month = refund.DeliveryMonth;
+            var year = refund.DeliveryYear;
+
+            while (amountToRefund < amountRefunded)
+            {
+                var stillToRefund = amountToRefund - amountRefunded;
+                var paymentsForPeriod = previousPayments
+                    .Where(x => x.DeliveryMonth == month && x.DeliveryYear == year)
+                    .ToList();
+                var newRefunds = ProcessRefundForPeriod(stillToRefund, year, month, paymentsForPeriod, refund);
+                amountRefunded += newRefunds.Sum(x => x.Amount);
+                refundPayments.AddRange(newRefunds);
+            
+                if (month == 8)
```

Comment:
This type of logic exist all over the system, to me it would make a lot of sense to turn this type of logic into a type and overload the + and - operators (but probably not for now). It would need to be in shared library